<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [151]:
import pandas as pd
import psycopg2

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [154]:
# текст запроса
n=10
query_3_1 = f'''select Count(*)
           from vacancies
           limit {n}
'''

In [155]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [156]:
query_3_2 = f'''
select count(*)
from employers
'''

In [157]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [158]:
query_3_3 = f'''
select count(*)
from areas
'''

In [159]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [160]:
query_3_4 = f'''
select count(*)
from industries
'''

In [161]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

База данных, предназначенная для работы, состоит из 5 таблиц, связанных ключами.
Самая большая таблица vacancies имеет около 50 тыс. строк, таблица employers из примерно 23 тыс. строк, далее идут более мелкие таблицы - areas, industries.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [162]:
query_4_1 = f'''
select a.name, count(*)
from vacancies v join areas a on v.area_id=a.id
group by a.name
order by count(*) desc
'''

In [163]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [164]:
query_4_2 = f'''
select count(*)
from vacancies
where salary_from is not null or salary_to is not null
'''

In [165]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [166]:
query_4_3 = f'''
select round(avg(salary_from)) as min_salary, round(avg(salary_to)) as max_salary
from vacancies
'''

In [167]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


min_salary  max_salary
0     71065.0    110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [168]:
query_4_4 = f'''
select schedule, employment, count(*)
from vacancies
group by schedule, employment
order by count(*) desc
'''

In [169]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [170]:
query_4_5 = f'''
select experience, count(*)
from vacancies
group by experience
order by count(*)
'''

In [171]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

1. Самое большое количество вакансий находится в крупнейших городах - Москва, Санкт-Петербург, Новосибирск, МинскЮ, Алматы
2. Мы видим большое количество пропусков в столбцах, посвященных заработной плате. В около 50% строк есть хотя бы 1 пропуск.
3. Средние значения для нижней и верхней границы зарплатной вилки составляют от 71065 до 110537 руб. 
4. Из анализа графика и типа трудоустройства можно однозначно сказать, что больше всего компании хотят, чтобы люди работали полную рабочую неделю.
5. Более 50% компаний готовы принимать людей с малым опытом работы (от 1 до 3 лет), что может говорить, что работадателям намного важнее знания рабочего, а не его многолетний опыт.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [172]:
query_5_1 = f'''
select e.name, count(*)
from vacancies v join employers e on v.employer_id=e.id
group by e.name
order by count(*) desc
'''

In [173]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0                      Яндекс   1933
1                  Ростелеком    491
2                    Тинькофф    444
3                        СБЕР    428
4               Газпром нефть    331
...                       ...    ...
14761            им. Тьюринга      1
14762        LLC Novalab Tech      1
14763   Компания Мир упаковки      1
14764  Лифинцев Олег Иванович      1
14765      ТрансЭнергоХолдинг      1

[14766 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [174]:
query_5_2 = f'''
select a.name as region, count(distinct e.id) as count_employers
from areas a left join employers e on a.id = e.area
where a.id not in (select distinct area_id from vacancies where area_id is not null)
group by a.name
order by count_employers desc
'''

In [175]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


region  count_employers
0                            Россия              410
1                         Казахстан              207
2                Московская область               75
3                Краснодарский край               19
4                          Беларусь               18
..                              ...              ...
588                        Кувандык                0
589                        Куйбышев                0
590  Куйбышево (Ростовская область)                0
591                        Култаево                0
592                        Кульсары                0

[593 rows x 2 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [176]:
query_5_3 = f'''
select e.name as name_of_employers, count(distinct v.area_id)
from employers e 
join vacancies v on e.id=v.employer_id
join areas a on e.area=a.id
group by e.name
order by count(distinct v.area_id) desc
'''

In [177]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name_of_employers  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [178]:
query_5_4 = f'''
select count(e.name)
from employers e 
left join employers_industries ei on e.id=ei.employer_id
left join industries i on ei.industry_id=i.id
where i.name is null

'''

In [179]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [180]:
query_5_5 = f'''
select e.name, count(i.id)
from employers e 
left join employers_industries ei on e.id=ei.employer_id
left join industries i on ei.industry_id=i.id
group by e.name
having count(i.id)=4
order by e.name
'''

In [181]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0        101 Интернет      4
1            21vek.by      4
2                2ГИС      4
3                  2К      4
4         4 пикселя +      4
...               ...    ...
1131            ЮРИОН      4
1132  ЮТИП Технологии      4
1133           ЯКласс      4
1134            ЯрНео      4
1135    Яшин&Партнёры      4

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [182]:
query_5_6 = f'''
select count(e.name)
from employers e 
left join employers_industries ei on e.id=ei.employer_id
left join industries i on ei.industry_id=i.id
where i.name like 'Разработка программного обеспечения'
'''

In [183]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [184]:
import requests
from bs4 import BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'

# Добавляем заголовки для имитации браузера
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
    'Accept-Language': 'ru-RU,ru;q=0.8,en-US;q=0.5,en;q=0.3',
    'Accept-Encoding': 'gzip, deflate',
    'Connection': 'keep-alive',
    'Upgrade-Insecure-Requests': '1',
}

response = requests.get(url, headers=headers)
page = BeautifulSoup(response.text, 'html.parser')

table = page.find('table') # считываем таблицу
cities = []
for row in table.find_all('tr')[1:]:
    cells = row.find_all('td')
    if cells: 
        city = cells[1].get_text(strip=True)
        cities.append(city)
print(cities)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [185]:
cities_sql = "('" + "', '".join(cities) + "')"

In [186]:
query_5_7 = f'''
select a.name, count(*) as count
from vacancies v 
join employers e on e.id=v.employer_id
join areas a on a.id=v.area_id
where e.name='Яндекс' and a.name in {cities_sql}
group by a.name

union all 

select 'Total', count(*) as count
from vacancies v 
join employers e on e.id=v.employer_id
join areas a on a.id=v.area_id
where e.name='Яндекс' and a.name in {cities_sql}

order by count;
'''

In [187]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

1. Самое большое количество вакансий наблюдается у Яндекса(1933), далее идет Ростелеком(491), Тинькофф(444) и СБЕР(428), пятёрку лидеров закрывает компания Газпром нефть с 331 вакансией.

2. Яндекс, Ростелеком и Спецремонт имеют широкую географию вакансий, так как ищут людей в более чем 100 регионах.

3. У большого количества работадателей (8419) не указана сфера деятельности, что, во-первых, снижает вероятность поиска этих компаний по фильтру, а, во-вторых, мешает в разработке модели машинного обучения, так как такие компании сложно будет использовать в дальнейшей работе.

4. Более 3 тыс. компаний указали в графе сфера деятельности - Разработка программного обеспечения

5. Компания Яндекс ищет сотрудников не только в городах-миллионниках, но и в более маленьких

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [188]:
query_6_1 = f'''
select count(*)
from vacancies
where lower(name) like '%data%' or lower(name) like '%данн%'
'''

In [189]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [190]:
query_6_2 = f'''
select count(*)
from vacancies
where (
  lower(name) like '%data scientist%' or
  lower(name) like '%data science%' or
  lower(name) like '%исследователь данных%' or
  lower(name) like '%machine learning%' or
  lower(name) like '%машинн%обучен%' or
  lower(name) like '%ml%' and lower(name) not like '%html%'
)
and (
  lower(name) like '%junior%' or
  lower(experience) = 'нет опыта' or
  lower(employment) = 'стажировка'
)

'''

In [191]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [192]:
query_6_3 = f'''
select count(*)
from vacancies
where (
  lower(name) like '%data scientist%' or
  lower(name) like '%data science%' or
  lower(name) like '%исследователь данных%' or
  lower(name) like '%machine learning%' or
  lower(name) like '%машинн%обучен%' or
  lower(name) like '%ml%' and lower(name) not like '%html%'
)
and (
  lower(key_skills) like '%sql%' or
  lower(key_skills) like '%postgres%'
)

'''

In [193]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [194]:
query_6_4 = f'''
select count(*)
from vacancies
where (
  lower(name) like '%data scientist%' or
  lower(name) like '%data science%' or
  lower(name) like '%исследователь данных%' or
  lower(name) like '%machine learning%' or
  lower(name) like '%машинн%обучен%' or
  lower(name) like '%ml%' and lower(name) not like '%html%'
)
and (
  lower(key_skills) like '%python%'
)

'''

In [195]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [196]:
query_6_5 = '''SELECT 
ROUND(AVG(CHAR_LENGTH(key_skills) - CHAR_LENGTH(replace(key_skills, CHR(9), '')) + 1), 2) AS avg_skills_count

from vacancies
where
    (lower(name) like '%data scientist%' or
    lower(name) like '%data science%' or
    lower(name) like '%исследователь данных%' or
    (name like '%ML%' and lower(name) not like'%html%') or
    lower(name) like '%machine learning%'or
    lower(name) like '%машинн%обучен%') and
    (key_skills is not null and key_skills != '')
'''

In [197]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\390248827.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


avg_skills_count
0              6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [198]:
query_6_6 = '''
select distinct experience,
       round(avg(coalesce(((salary_from + salary_to) / 2.0), salary_from, salary_to))) as avg_salary
from vacancies
where
    (lower(name) like '%data scientist%' or
    lower(name) like '%data science%' or
    lower(name) like '%исследователь данных%' or
    (name like '%ML%' and lower(name) not like'%html%') or
    lower(name) like '%machine learning%'or
    lower(name) like '%машинн%обучен%') and
    (salary_from is not null or salary_to is not null)
group by experience
'''

In [199]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

1. Только 1771 вакансий имеют отношение к данным, хотя их общее количество около 50 000

2. junior дата-сайентистов в таблице ещё меньше, всего 51 вакансия.

3. 229 вакансий в таблице в качестве ключевого навыка выделяют знание SQL или postgres

4. Python является более популярным, вакансий с ключевым навыком python - 357

5. Среднее количество ключевых навыков для дата-саейнтистов равно 6.41

6. Дата-сайентист с опытом работы от 3 до 6 лет может в среднем рассчитывать на заработную плату в размере 243 тыс. руб.

___


# Самостоятельные исследования:
* 1. Топ-10 вакансий с самой высокой верхней границей зарплаты.

* 2. Средняя зарплата (по mid-уровню: (salary_from + salary_to)/2) по типам занятости (employment).

* 3. Вакансии с самой широкой вилкой (разница между salary_to и salary_from).

* 4. Доля вакансий с указанной зарплатой (где заполнены оба поля salary_from и salary_to).

* 5. Топ-5 регионов по средней предлагаемой зарплате.

1. Топ-10 вакансий с самой высокой верхней границей зарплаты. 

In [200]:
query_1 = '''
select name, key_skills, experience, employment, salary_to
from vacancies
where salary_to is not null
order by salary_to desc
limit 10
'''

In [201]:
df = pd.read_sql_query(query_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\3139391625.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_1, connection)


name  \
0                             QA Automation Engineer   
1                         СТО/Директор по разработке   
2                     Брокер по продаже апартаментов   
3                1С:Эксперт по производительности 1С   
4  Senior Software Developer (upwork / иностранны...   
5                Senior DevOps in Blockchain Startup   
6             FullStack Developer in Unicorn Company   
7              Frontend Developer in Unicorn Company   
8                            Product manager LATOKEN   
9                            Product manager LATOKEN   

                                          key_skills          experience  \
0  Тестирование\tHTML\tJavaScript\tGit\tТестирова...  От 1 года до 3 лет   
1  Laravel\tУправление проектами\tPHP\tVue.js\tPo...       От 3 до 6 лет   
2  Ведение переговоров\tТелефонные переговоры\tРа...  От 1 года до 3 лет   
3     PostgreSQL\tMS SQL Server\t1С программирование       От 3 до 6 лет   
4  Английский язык\tДеловое общение\tУдаленная ра...       От 3 до 6 лет   
5  AWS\tGCP\tDevOps\tGo\tGolang\tPython\tDocker\t...       От 3 до 6 лет   
6  Java\tReact\tKotlin\tSpring Boot\tAWS\tGCP\tMo...       От 3 до 6 лет   
7  Node.js\tJavaScript\tMongoDB\tAngular\tSpring ...       От 3 до 6 лет   
8  Product Management\tBusiness Development\tTeam...       От 3 до 6 лет   
9  Product Management\tBusiness Development\tTeam...       От 3 до 6 лет   

            employment  salary_to  
0     Полная занятость    1000000  
1     Полная занятость    1000000  
2     Полная занятость    1000000  
3     Полная занятость     950000  
4  Частичная занятость     935176  
5     Полная занятость     860000  
6     Полная занятость     860000  
7     Полная занятость     860000  
8     Полная занятость     800000  
9     Полная занятость     800000

2. Средняя зарплата (по mid-уровню: (salary_from + salary_to)/2) по типам занятости (employment)

In [202]:
query_2 = '''
select distinct employment,
       round(avg(coalesce(((salary_from + salary_to) / 2.0), salary_from, salary_to))) as avg_salary
from vacancies
group by employment
'''

In [203]:
df = pd.read_sql_query(query_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\4279923753.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_2, connection)


employment  avg_salary
0     Полная занятость     88240.0
1     Проектная работа     92086.0
2           Стажировка     37249.0
3  Частичная занятость     31566.0

3. Вакансии с самой широкой вилкой (разница между salary_to и salary_from).

In [204]:
query_3 = '''
select name, key_skills, experience, employment, salary_from, salary_to, (salary_to - salary_from) as salary_range
from vacancies
where salary_to is not null and salary_from is not null
order by salary_range desc
limit 10
'''

In [205]:
df = pd.read_sql_query(query_3, connection)
df


C:\Users\user\AppData\Local\Temp\ipykernel_1764\1309112477.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3, connection)


name  \
0  Senior Software Developer (upwork / иностранны...   
1                1С:Эксперт по производительности 1С   
2                             QA Automation Engineer   
3                    Программист-разработчик C#, SQL   
4                    Программист-разработчик C#, SQL   
5                Senior DevOps in Blockchain Startup   
6                         СТО/Директор по разработке   
7                     Брокер по продаже апартаментов   
8                       ML Engineer \ Data scientist   
9                                Специалист по пайке   

                                          key_skills          experience  \
0  Английский язык\tДеловое общение\tУдаленная ра...       От 3 до 6 лет   
1     PostgreSQL\tMS SQL Server\t1С программирование       От 3 до 6 лет   
2  Тестирование\tHTML\tJavaScript\tGit\tТестирова...  От 1 года до 3 лет   
3  C#\tASP.NET\tWPF\t.NET Framework\tJavaScript\t...  От 1 года до 3 лет   
4  C#\tASP.NET\tWPF\t.NET Framework\tJavaScript\t...  От 1 года до 3 лет   
5  AWS\tGCP\tDevOps\tGo\tGolang\tPython\tDocker\t...       От 3 до 6 лет   
6  Laravel\tУправление проектами\tPHP\tVue.js\tPo...       От 3 до 6 лет   
7  Ведение переговоров\tТелефонные переговоры\tРа...  От 1 года до 3 лет   
8  Нейронные сети\tУдаленная работа\tMachine Lear...       От 3 до 6 лет   
9  Работа в команде\tРабота в условиях многозадач...       От 3 до 6 лет   

            employment  salary_from  salary_to  salary_range  
0  Частичная занятость       175346     935176        759830  
1     Полная занятость       350000     950000        600000  
2     Полная занятость       400000    1000000        600000  
3     Полная занятость        40000     600000        560000  
4     Полная занятость        40000     600000        560000  
5     Полная занятость       315000     860000        545000  
6     Полная занятость       500000    1000000        500000  
7     Полная занятость       500000    1000000        500000  
8     Полная занятость       200000     700000        500000  
9     Полная занятость        50000     500000        450000

4. Доля вакансий с указанной зарплатой (где заполнены оба поля salary_from и salary_to).

In [206]:
query_4 = '''
select round((select count(*) 
        from vacancies 
        where salary_from is not null and salary_to is not null) * 100.0 / count(*), 3)
from vacancies 
'''

In [207]:
df = pd.read_sql_query(query_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\1005795368.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4, connection)


round
0  20.534

5. Топ-5 регионов по средней предлагаемой зарплате.

In [208]:
query_5 = '''
select a.name, round(avg(coalesce(((salary_from + salary_to) / 2.0), salary_from, salary_to))) as avg_salary
from vacancies v join areas a on v.area_id=a.id
where salary_to is not null or salary_from is not null
group by a.name
order by avg_salary desc
limit 5
'''

In [209]:
df = pd.read_sql_query(query_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_1764\3874968683.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5, connection)


name  avg_salary
0    Германия    452751.0
1    Болгария    324348.0
2     Испания    300000.0
3  Черногория    288589.0
4       Литва    263019.0

# Выводы
1. исследуя названия вакансий и ключевые навыки самых высокооплачиваемых профессий, можно выделить ряд навыков, которые больше других ценятся на рынке труда

2. Самая высокая средняя заработная плата наблюдается у людей на проектной работе и полной занятости

3. Высокая зарплатная вилка даёт много неопределенности как для работодателя, так и для модели машинного обучения, поэтому с такими вакансиями следует работать намного внимательнее.

4. В таблице только 20 % вакансий имеют как верхний, так и нижний уровень заработной платы. Данный факт существенно усложняет работу с этими признаками. 

5. Самые высокие заработные платы наблюдаются в странах европы.


# Общий вывод по проекту

Представленные таблицы содержат все необходимые признаки для полноценного анализа вакансий, включая информацию о опыте, требованиях, уровне зарплаты и местоположении. Это позволяет выявлять важные закономерности рынка труда, определять наиболее востребованные навыки и оценивать предложения работодателей в разных регионах. На основе этих данных можно проводить детальные исследования и формировать обоснованные выводы о текущих трендах. Кроме того, структура и полнота информации делают таблицы отличной основой для создания моделей машинного обучения. 
